src/ui/public/share/directives/share.js

```diff
@@ -1,18 +1,109 @@
+import {
+  getUnhashableStatesProvider,
+  unhashUrl,
+} from 'ui/state_management/state_hashing';
+
+import urlShortenerProvider from '../lib/url_shortener';
+
 import uiModules from 'ui/modules';
 import shareTemplate from 'ui/share/views/share.html';
 const app = uiModules.get('kibana');
 
-app.directive('share', function () {
+app.directive('share', function (Private, Notifier) {
+  const getUnhashableStates = Private(getUnhashableStatesProvider);
+  const urlShortener = Private(urlShortenerProvider);
+
   return {
     restrict: 'E',
     scope: {
       objectType: '@',
       objectId: '@',
-      setAllowEmbed: '&?allowEmbed'
+      allowEmbed: '@allowEmbed'
     },
     template: shareTemplate,
-    controller: function ($scope) {
-      $scope.allowEmbed = $scope.setAllowEmbed ? $scope.setAllowEmbed() : true;
+    controllerAs: 'share',
+    controller: function ($scope, $location, globalState) {
+      // Default to allowing an embedded IFRAME, unless it's explicitly set to false.
+      this.allowEmbed = $scope.allowEmbed === 'false' ? false : true;
+      this.objectType = $scope.objectType;
+
+      function getOriginalUrl(url) {
+        // If there is no objectId, then it isn't saved, so it has no original URL.
+        if ($scope.objectId === undefined || $scope.objectId === '') {
+          return;
+        }
+
+        // Remove the query string.
+        return url.split('?')[0];
+      };
+
+      function getSnapshotUrl(url) {
+        // Replace hashes with original RISON values.
+        const urlWithStates = unhashUrl(url, getUnhashableStates());
+        return urlWithStates;
+      }
+
+      this.getEmbeddableUrl = url => {
+        const embedQueryParam = '?embed=true';
+        const urlHasQueryString = url.indexOf('?') !== -1;
+        const embeddableUrl =
+          urlHasQueryString ? url.replace('?', `${embedQueryParam}&`) : `${url}${embedQueryParam}`;
+        return embeddableUrl;
+      };
+
+      this.getIframeUrl = url => {
+        const embeddableUrl = this.getEmbeddableUrl(url);
+        return `<iframe src="${embeddableUrl}" height="600" width="800"></iframe>`;
+      };
+
+      // Original URLs.
+      this.originalUrl;
```

Comment:
How would you feel about putting all the urls under `this.urls = { original, snapshot, ... }`. Then, in `updateUrls()` you can just overwrite the `this.urls` object in order to deal with https://github.com/elastic/kibana/pull/8172#issuecomment-245335433